appgate/resource_appgate_appliance_controller.go

```diff
@@ -0,0 +1,253 @@
+		DeleteContext: resourceAppgateApplianceControllerActivationDelete,
+
+		Timeouts: &schema.ResourceTimeout{
+			Create: schema.DefaultTimeout(15 * time.Minute),
+			Update: schema.DefaultTimeout(15 * time.Minute),
+			Delete: schema.DefaultTimeout(15 * time.Minute),
+		},
+		Schema: func() map[string]*schema.Schema {
+			s := map[string]*schema.Schema{
+				"appliance_id": {
+					Type:        schema.TypeString,
+					Description: "ID of the object.",
+					Required:    true,
+				},
+				"controller":      controllerSchema(),
+				"admin_interface": adminInterfaceSchema(),
+			}
+			// all these fields are mandatory within this resource
+			s["admin_interface"].Optional = false
+			s["admin_interface"].Required = true
+			s["controller"].Optional = false
+			s["controller"].Required = true
+			s["controller"].Computed = false
+			return s
+		}(),
+	}
+}
+
+func resourceAppgateApplianceControllerActivationCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	// Warning or errors can be collected in a slice type  var diags diag.Diagnostics
+	var diags diag.Diagnostics
+	Appliance54Constraints, err := version.NewConstraint(">= 5.4.0")
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	token, err := meta.(*Client).GetToken()
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	currentVersion := meta.(*Client).ApplianceVersion
+	if !Appliance54Constraints.Check(currentVersion) {
+		return diag.FromErr(errors.New("This resource is only avaliable in 5.4 or higher."))
+	}
+
+	id := d.Get("appliance_id").(string)
+	log.Printf("[DEBUG] Creating Appliance Controller activation with name: %s", id)
+	api := meta.(*Client).API.AppliancesApi
+	request := api.AppliancesIdGet(ctx, id)
+	appliance, res, err := request.Authorization(token).Execute()
+	if err != nil {
+		d.SetId("")
+		if res.StatusCode == http.StatusNotFound {
+			return nil
+		}
+		return diag.FromErr(fmt.Errorf("Failed to read Appliance, %+v", err))
+	}
+
+	if v := appliance.GetActivated(); !v {
+		return diag.FromErr(fmt.Errorf("Can not activate controller functions on an inactive appliance. The appliance %q need to be seeded first.", appliance.GetName()))
+	}
+	if v, ok := d.GetOk("controller"); ok {
+		ctrl, err := readControllerFromConfig(v.([]interface{}))
+		if err != nil {
+			return diag.FromErr(err)
+		}
+		appliance.SetController(ctrl)
+	}
+	if a, ok := d.GetOk("admin_interface"); ok {
+		ainterface, err := readAdminInterfaceFromConfig(a.([]interface{}))
+		if err != nil {
+			return diag.FromErr(err)
+		}
+		appliance.SetAdminInterface(ainterface)
+	}
+
+	retryErr := resource.Retry(d.Timeout(schema.TimeoutCreate), func() *resource.RetryError {
+		_, _, err := api.AppliancesIdPut(ctx, id).Appliance(appliance).Authorization(token).Execute()
+		if err != nil {
+			return resource.NonRetryableError(err)
+		}
+		// wait a few seconds and make sure the controller react to the changes.
+		time.Sleep(time.Duration(10) * time.Second)
```

Comment:
Is a sleep really required here? Can't `waitForController` be able to handle it regardless?